server/genquery2/src/genquery2_sql.cpp

```diff
@@ -0,0 +1,1172 @@
+                _state.sql_tables.push_back(_column_info.table);
+                _state.table_aliases[std::string{_column_info.table}] = generate_table_alias(_state);
+            }
+        }
+
+        return {is_special_column, table_alias};
+    } // setup_column_for_post_processing
+
+    auto to_sql(gq_state& _state, const column& _column) -> std::string
+    {
+        const auto iter = column_name_mappings.find(_column.name);
+
+        if (iter == std::end(column_name_mappings)) {
+            throw std::invalid_argument{fmt::format("unknown column: {}", _column.name)};
+        }
+
+        // Capture all column objects as some parts of the implementation need to access them in
+        // order to generate the proper SQL.
+        _state.ast_column_ptrs.push_back(&_column);
+
+        auto [is_special_column, table_alias] = setup_column_for_post_processing(_state, _column, iter->second);
+        const std::string_view alias =
+            is_special_column ? table_alias : _state.table_aliases.at(std::string{iter->second.table});
+
+        if (_column.type_name.empty()) {
+            return fmt::format("{}.{}", alias, iter->second.name);
+        }
+
+        return fmt::format("cast({}.{} as {})", alias, iter->second.name, _column.type_name);
+    }
+
+    auto to_sql(gq_state& _state, const select_function& _select_function) -> std::string
+    {
+        const auto iter = column_name_mappings.find(_select_function.column.name);
+
+        if (iter == std::end(column_name_mappings)) {
+            throw std::invalid_argument{fmt::format("unknown column: {}", _select_function.column.name)};
+        }
+
+        // Capture all column objects as some parts of the implementation need to access them later in
+        // order to generate the proper SQL.
+        _state.ast_column_ptrs.push_back(&_select_function.column);
+
+        // Aggregate functions are not allowed in the WHERE clause of an SQL statement!
+        if (!_state.in_select_clause) {
+            throw std::invalid_argument{"aggregate functions not allowed in where clause"};
+        }
+
+        auto [is_special_column, table_alias] =
+            setup_column_for_post_processing(_state, _select_function.column, iter->second);
+        const std::string_view alias =
+            is_special_column ? table_alias : _state.table_aliases.at(std::string{iter->second.table});
+
+        if (_select_function.column.type_name.empty()) {
+            return fmt::format("{}({}.{})", _select_function.name, alias, iter->second.name);
+        }
+
+        return fmt::format("{}(cast({}.{} as {}))",
+                           _select_function.name,
+                           alias,
+                           iter->second.name,
+                           _select_function.column.type_name);
+    }
+
+    auto to_sql(gq_state& _state, const selections& _selections) -> std::string
+    {
+        irods::at_scope_exit restore_value{[&_state] { _state.in_select_clause = false; }};
```

Comment:
For my own understanding... Is it possible for `in_select_clause` to be `true` at the start? Seems like probably not, but this is less "restoring" the value as it is setting it to false.

Reply:
Not in the current implementation.

Every call to the parser results in a newly initialized state object. `in_select_clause` may need to be revisited once we start thinking about sub-selects. Who knows when that'll happen though.

Reply:
Okay. Wasn't sure if it would be worth capturing the value of `_state.in_select_clause` at the top here and then restoring the value to that, rather than always to `false`. Maybe not.

Reply:
Okey doke.